libraries/reconstruction/ctf_estimate_from_psd_fast.cpp

```diff
@@ -1124,38 +1147,94 @@ void ProgCTFEstimateFromPSDFast::estimate_defoci_fast()
 			double widx=w/deltaW;
 			size_t lowerIdx=floor(widx);
 			double weight=widx-floor(widx);
-			double value =(1-weight)*A1D_ELEM(psd_background,lowerIdx)
-												 +weight*A1D_ELEM(psd_background,lowerIdx+1);
+			if(i < XSIZE(psd_exp_radial2)-1)
+			{
+				double value =(1-weight)*A1D_ELEM(psd_background,lowerIdx)
+													 +weight*A1D_ELEM(psd_background,lowerIdx+1);
 
-			A1D_ELEM(psd_exp_radial2,i) = value*value;
+				A1D_ELEM(psd_exp_radial2,i) = value*value;
+			}
 		}
-
+#ifdef DEBUG
+		std::cout << "psd_exp_radial2\n" << psd_exp_radial2 << std::endl;
+		std::cout << "--------------------------------------" << std::endl;
+#endif
 		FourierTransformer FourierPSD;
 		FourierPSD.FourierTransform(psd_exp_radial2, psd_fft, false);
 
-		int startIndex = 7;
+		const double minDefocus=2000;
+		int startIndex = ceil(std::max((double)7,minDefocus*current_ctfmodel.lambda/(2*pow(2*current_ctfmodel.Tm,2))));
+#ifdef DEBUG
+		std::cout << "Start index=" << minDefocus*current_ctfmodel.lambda/(2*pow(2*current_ctfmodel.Tm,2)) << std::endl;
+		std::cout << "--------------------------------------" << std::endl;
+#endif
 		if (current_ctfmodel.VPP_radius != 0.0)
 			startIndex = 10; //avoid low frequencies
 		FOR_ALL_ELEMENTS_IN_ARRAY1D(psd_fft)
 		{
-			if(i>=startIndex)
+			if(i>=startIndex) {
 				amplitud.push_back(abs(psd_fft[i]));
-
+#ifdef DEBUG
+		std::cout << abs(psd_fft[i]) << std::endl;
+#endif
+			}
 		}
+#ifdef DEBUG
+		std::cout << "--------------------------------------" << std::endl;
+#endif
 
-		double maxValue = *max_element(amplitud.rbegin(),amplitud.rend());
-		int finalIndex = 0;
+		//Remove outliers
+#ifdef OUTLIERS
+		double totalSum = 0;
+		for(size_t i=0;i<amplitud.size();i++){
+			totalSum += amplitud[i];
+		}
+		double amplitudMean = totalSum / amplitud.size();
+		double sdSum = 0;
 		for(size_t i=0;i<amplitud.size();i++)
 		{
-			if(maxValue == amplitud[i])
+			double diff=amplitud[i]-amplitudMean;
+			sdSum += diff*diff;
+		}
+		double amplitudSD = sqrt(sdSum/amplitud.size());
+		double differenceSD = 3*amplitudSD;
+		for(size_t i=0;i<amplitud.size();i++){
+			if (amplitud[i]>amplitudMean+differenceSD){
+				amplitud[i] = amplitudMean+differenceSD;
+			}
+		}
+#endif
+
+		double maxValue=-1e38;
+		int finalIndex=-1;
+#ifdef DEBUG
+		std::cout << "Looking for maximum ...\n";
+#endif
+		for (size_t i=1; i<amplitud.size()-1; i++)
+		{
+#ifdef DEBUG
+			std::cout << "i=" << i << " amplitude i= " << amplitud[i] << std::endl;
+#endif
+			if (amplitud[i]>maxValue && amplitud[i]>amplitud[i-1] && amplitud[i]>=amplitud[i+1])
```

Comment:
First condition implies second condition (i.e. will always be true).
Why do you have so 'complicated' check anyway?